dyno: resolve param for loops

benharsh opened the pull request:
Resolves https://github.com/Cray/chapel-private/issues/3393

This PR implements resolution of param for-loops, the results of which will be stored in the new type ``ResolvedParamLoop``.

We resolve param loops by creating a new Resolver for each iteration, and
storing the collected ResolutionResultByPostorderIDs in a
ResolvedParamLoop result. Each iteration stores a different param
resolution result for the AST representing the index variable.

These temporary Resolvers don't inherit their parent's scopeStacks or
their ResolutionResultByPostorderID values directly. Instead, a
'parentResolver' reference is stored in the Resolver type and used by
'typeForId' to determine whether a parent resolver 'owns' the resolution
results for the desired scope.

The ResolvedParamLoop is finally stored in the ResolvedExpression for
the For loop.

-----

This PR introduces a first draft of a ResolvedVisitor type, which is meant
to handle iterating over the AST and make resolution results available to
the client type.

Clients of this type will implement their own visitors, e.g.
MyVisitor, which will implement enter/exit methods that accept two
arguments:
- An AST node
- A reference to a ResolvedVisitor<MyVisitor>

The ResolvedVisitor stores a ResolutionResultByPostorderID that will
contain the relevant results for the corresponding AST.

The ResolvedVisitor will handle param for loops by first invoking the
client 'enter' method to determine whether it should recurse. If so, it
will traverse once over the iterand, and then will traverse for each
iteration of the loop body with the corresponding results from the
ResolvedParamLoop.

Remaining work:
- handle strided ranges
- handle ranges with '#' or '<' operators

Testing:
- [x] test-dyno

mppf commented on compiler/dyno/include/chpl/resolution/resolution-types.h:
```diff
@@ -29,6 +29,7 @@
 #include "chpl/uast/Function.h"
 #include "chpl/util/bitmap.h"
 #include "chpl/util/memory.h"
+#include "chpl/uast/all-uast.h"
```
Can we forward declare instead? Like



```
namespace chpl {
  namespace uast {
    class For;
  }
}
```



That's the strategy I've been trying to use elsewhere, just to try to keep the amount of C++ headers required (and compile times) down.

mppf commented on compiler/dyno/include/chpl/resolution/resolution-types.h:
```diff
@@ -1380,6 +1396,51 @@ class ResolvedFields {
+    std::vector<ResolutionResultByPostorderID> loopBodies_;
+
+  public:
+    using LoopBodies = std::vector<ResolutionResultByPostorderID>;
```
IMO it would be nice to put this in its own `public:` block just before the private field decls, but it's not a big deal either way.

mppf commented on compiler/dyno/lib/resolution/Resolver.cpp:
```diff
@@ -1841,67 +1868,111 @@ void Resolver::exit(const uast::New* node) {
+
+        cur.exitScope(loop);
+
+        loopResults.push_back(cur.byPostorder);
```
This might do an expensive copy since byPostorder might be a vector and all that. The main way I know of to avoid the copy is `loopResults.push_back(ResolutionResultByPostorderID()); loopResults.back().swap(cur.byPostorder)`. But you might know of a better way.

DanilaFe replied:
Will `std::move`'ing the `byPostorder` not work? `push_back` has an rvalue reference overload.

mppf replied:
Seems like a good plan if it would work! I keep thinking that `push_back` only takes in a reference so it wouldn't work but I'm always forgetting / confusing myself about rvalue references.

mppf reviewed (approved):
Cool! One thing I am curious about is with the ResolvedVisitor, why does it pass a reference to the ResolvedVisitor to the enter/exit functions. I would think that passing a `const ResolvedExpression&` would make more sense. But maybe there is some case you are thinking of.

benharsh commented:
@mppf : I believe I've addressed your comments.

These methods accepted a ``ResolvedVisitor`` because if a UserVisitor ever wanted to manually traverse an AstNode then they should do so through the ``ResolvedVisitor``. The current, limited, use of this type in testLoopIndexVars doesn't need to do that sort of manual traversal. Something else, like a typed AST printer, might want to do that someday.
